Add the default enrollment start date on course creation

dyudyunov opened the pull request:
### Problem
Course is visible in the main page right after creation. So anonymous users can see them and access course about page for the courses without valid data (e.g. they will see default course overview)

### Root cause
When courses list filtering processed it checks the `see_exists` permission for the anonymous user.
Actually, `see_exists` means `can_load` OR `can_enroll`.

`can_load` is False in our case because course start is in the future.

But `can_enroll` returns True because course's enrollment_start and enrollment_end dates are blank:
```
enrollment_start = courselike.enrollment_start or datetime.min.replace(tzinfo=UTC)
enrollment_end = courselike.enrollment_end or datetime.max.replace(tzinfo=UTC)
if enrollment_start < now < enrollment_end:
    debug("Allow: in enrollment period")
    return ACCESS_GRANTED
```

### Fix
Set the enrollment_start the same as a course start by default

### Known Issues:
- this requires the https://github.com/openedx/openedx-demo-course/pull/38 to be merged. Otherwise, it will break the clean (first) deployment while trying to enroll `audit` and `verified` users in the Demo Course

### Notes
- this fix will work for the course catalog page when the course discovery is disabled (FEATURES['ENABLE_COURSE_DISCOVERY'] = False)
- I wonder why DEFAULT_START_DATE is hardcoded (1 Jan 2030)? How and when is it updated?
- The proposed behavior is togglable and is disabled by default. To enable it - update your settings.FEATURES for cms:
`FEATURES['CREATE_COURSE_WITH_DEFAULT_ENROLLMENT_START_DATE'] = True`

openedx-webhooks commented:
Thanks for the pull request, @dyudyunov!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.




<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

dyudyunov committed:
fix: main page course listing

The course is visible on the main page right after creation.

So anonymous users can see them and access the course about page
for the courses without valid data (e.g. they will see the default
course overview)

When courses list filtering is processed it checks the `see_exists`
permission for the anonymous user.
Actually, `see_exists` means `can_load` OR `can_enroll`.

`can_load` is False in our case because the course start in the future.

But `can_enroll` returns True because the course's enrollment_start
and enrollment_end dates are blank:
```
enrollment_start = courselike.enrollment_start or datetime.min.replace(tzinfo=UTC)
enrollment_end = courselike.enrollment_end or datetime.max.replace(tzinfo=UTC)
if enrollment_start < now < enrollment_end:
    debug("Allow: in enrollment period")
    return ACCESS_GRANTED
```

Set the enrollment_start the same as a course start by default

- this fix will work for the course catalog page when the course
discovery is disabled (FEATURES['ENABLE_COURSE_DISCOVERY'] = False)
- I wonder why DEFAULT_START_DATE is hardcoded (1 Jan 2030)?
How and when is it updated?

dyudyunov commented:
@natabene this one is ready for the review

nedbat commented:
📣 💥 Heads-up: You must either **rebase onto master** or **merge master into your branch** to keep passing required checks.

We added a new required check, "Tests Successful," that this PR does not yet run.  Rebasing will get it started.

If you have any questions, please reach out to the Architecture team (either [#architecture](https://openedx.slack.com/archives/C024JPE27PV) on Open edX Slack or [#external-architecture](https://edx-internal.slack.com/archives/C024JPE27PV) on edX internal Slack).

dyudyunov commented:
@nedbat rebase is done

mariajgrimaldi commented:
I can take this one :)

mphilbrick211 commented:
Hi @dyudyunov! Your CLA check looks stuck - would you mind rebasing to see if that helps it go through?

Also, friendly ping @mariajgrimaldi for review :)

mariajgrimaldi commented on xmodule/tests/test_course_module.py:
```diff
@@ -32,10 +33,13 @@
+        assert xmodule.course_module.CourseFields.start.default == DEFAULT_START_DATE
+
+    def test_default_enrollment_start_date(self):
+        assert xmodule.course_module.CourseFields.start.default == DEFAULT_START_DATE
```
What's the difference between these two tests? maybe the second one should be `enrollment_start` instead of just `start`?

dyudyunov replied:
oh
yes, you're right, I'll fix this

mariajgrimaldi reviewed (commented):
This looks pretty good! I just left a question.

Also, I had the same question regarding the `DEFAULT_START_DATE` constant, and maybe [this](https://github.com/openedx/edx-platform/pull/15170#discussion_r118539083) is the reason. I was going to suggest changing it, but maybe that should be part of another effort.

mariajgrimaldi commented:
@mphilbrick211: thanks for the reminder :)

mariajgrimaldi commented:
I tested before and after this PR:

**Before**
![course-schedule-before-pr-30954](https://user-images.githubusercontent.com/64440265/208416620-8677070d-f724-44d9-aa6d-0633be6a50bc.png)


**After**
![course-schedule-after-pr-30954](https://user-images.githubusercontent.com/64440265/208416628-74744b75-513c-4f5a-bbcc-207067b8dda8.png)

The course was still visible when accessing the main listing page, but the enrollment was closed for an anonymous user. I first thought the course would be removed from the main page listing. Is this the behavior you expected? If yes, we should change the PR name since it suggests we're fixing the main page listing instead of the enrollment for brand-new courses.

dyudyunov commented:
@mariajgrimaldi no, it's not the expected behaviour 

Please, confirm that:
- your `COURSE_CATALOG_VISIBILITY_PERMISSION` setting is set to `see_exist` (the default one)
- you have cleared the cache after your change in course dates or removed the `@cache_if_anonymous()` decorator for the view

mariajgrimaldi commented:
@dyudyunov: it's not working for me, the course is still visible in the course listing. I'll be researching what's wrong with my setup.

dyudyunov commented:
@mariajgrimaldi 
try this way:
- create a brand new course
- check that it has the enrollment start by default
- check the `localhost:18000/` page to verify it is not shown there (I've tested with the anonymous user in the incognito window)

NOTE: the changes are aimed at the [`_can_enroll_courselike` method](https://github.com/openedx/edx-platform/blob/99cd4a4715f8dcd29b4cdd211a35cc02b5181a75/lms/djangoapps/courseware/access.py#L235). So checking it I supposed you're using the staff account or your user could be in the CourseEnrollmentAllowed table.

Also, I want to remind you that the https://github.com/openedx/openedx-demo-course/pull/38 should be merged first to avoid the first-deployment issues related to test users enrollments in the demo course (the `honor` user will not be able to enroll because enrollment_start is in the future)

mariajgrimaldi commented:
@dyudyunov: good! thanks for the guidance. It's working now.

@mphilbrick211: can we ping someone to review the dependent PR (https://github.com/openedx/openedx-demo-course/pull/38)? I can review but I don't have the right permissions for merging.

mphilbrick211 commented:
Thanks for flagging, @mariajgrimaldi! I have pinged Tim and Jeremy in #38 https://github.com/openedx/openedx-demo-course/pull/38

itsjeyd commented:
Hey @dyudyunov! While https://github.com/openedx/openedx-demo-course/pull/38 is waiting for review (left another [ping](https://github.com/openedx/openedx-demo-course/pull/38#issuecomment-1373701725) there just now), could you please fix merge conflicts here?

CC @mphilbrick211 @mariajgrimaldi

dyudyunov commented:
> Hey @dyudyunov! While [openedx/openedx-demo-course#38](https://github.com/openedx/openedx-demo-course/pull/38) is waiting for review (left another [ping](https://github.com/openedx/openedx-demo-course/pull/38#issuecomment-1373701725) there just now), could you please fix merge conflicts here?
> 
> CC @mphilbrick211 @mariajgrimaldi

Done

itsjeyd commented:
Thanks @dyudyunov!

I'm marking this as blocked on other work again -- technical review for https://github.com/openedx/openedx-demo-course/pull/38 is complete now (cf. https://github.com/openedx/openedx-demo-course/pull/38#pullrequestreview-1238989191) but that PR also [needs product review](https://github.com/openedx/openedx-demo-course/pull/38#issuecomment-1378477862).

CC @mphilbrick211 @mariajgrimaldi

mphilbrick211 commented:
@mariajgrimaldi - now that [#38](https://github.com/openedx/openedx-demo-course/pull/38#issuecomment-1422932383) is merged, would you mind reviewing/merging this one? It was blocked by #38.

mariajgrimaldi commented:
@mphilbrick211 of course! thanks for the reminder :)

mariajgrimaldi commented:
I've been thinking hard about this PR. Now, enrollment starts will have a default making courses unavailable for anonymous users after creation, which sounds good, IMO. My issue relies on the fact that this PR doesn't sound like a bug fix but a behavior change we need to study a bit more.

@mphilbrick211: can we get a hand from the product side? I'm not sure this new behavior is what we want for the platform.

mphilbrick211 commented:
@jmakowski1123 flagging this for you.

jmakowski1123 commented:
Hi @dyudyunov , Hi @mariajgrimaldi 

Thanks for this contribution! I agree with @mariajgrimaldi , that this PR is proposing a behavior change to the platform, rather than fixing a bug. There are more details I'd like to understand from a product perspective. 

Because this is a product-oriented proposal and the actual scope may be bigger than this single PR, I've created a Feature Ticket [here](https://github.com/openedx/platform-roadmap/issues/234). I'd like to shift the communication about product questions to this ticket. I wrote some questions in the comments that need some clarification. @dyudyunov , can we continue chatting [here](https://github.com/openedx/platform-roadmap/issues/234)?

mphilbrick211 commented:
@jmakowski1123 any update on this?

mphilbrick211 commented:
> @jmakowski1123 any update on this?

@jmakowski1123 - disregard this - I pinged Steve in the roadmap ticket.

mphilbrick211 commented:
Hi @mariajgrimaldi! Product review is complete on this - would you mind reviewing and merging if you're able? Thanks!

mariajgrimaldi commented:
I'll review this again today. Sorry for the delay

mphilbrick211 commented:
Hi @dyudyunov! We've had a few new tests pop up on this pull request (shellcheck). Would you mind running these? Thanks!

CC: @mariajgrimaldi

mariajgrimaldi commented:
In case anyone wants to replicate the testing of this PR, this is what I did. First, I configured my testing environment by doing the following:

1. Ensure FEATURES['ENABLE_COURSE_DISCOVERY'] = False
2. Ensure COURSE_CATALOG_VISIBILITY_PERMISSION = 'see_exists' (it's not the default in tutor dev)
3. Disable cache from course listing page
4. Create brand new course

Then, I tested two case scenarios:

**1. Courses with enrollment date in the future are not visible from the main course page**: after creating the course, check it has default enrollment date and it's not visible from the main course listing.

**2. Courses with enrollment date in the future are visible from the main course page**: after creating the course, check it has default enrollment date and it's visible from the main course listing. Besides the basic setup we did previously, for this test case I:

- Enabled course waffle flag: course_experience.pre_start_access
- Ensured visible_to_staff_only = False
- Ensured course is not expired

Which made the course visible although its enrollment date was far in the future.

mariajgrimaldi reviewed (approved):
@dyudyunov: this looks good! Thank you for being so patient. Can you change the PR name for something more descriptive of the changes?

mariajgrimaldi commented:
@jmakowski1123 @mphilbrick211: hello there! before merging this PR, should we notify someone of these changes?

dyudyunov commented:
> @dyudyunov: this looks good! Thank you for being so patient. Can you change the PR name for something more descriptive of the changes?

Hi @mariajgrimaldi 
Thanks for the precise review 💪

I've changed the title, please let me know if it's not descriptive enough

dyudyunov commented:
@mphilbrick211 hi

I've rebased on the master to trigger the tests run

dyudyunov commented:
some tests failed on the ubuntu packages fetching 🤔

mariajgrimaldi commented:
@jmakowski1123 @mphilbrick211: I asked internally about these changes and folks at 2U/edX don't want to implement them unless it is toggleable. [Here's ](https://openedx.slack.com/archives/C02QN50TYMD/p1682605354607819) the thread. Was this considered during the product review? Is this something we should include in the future? cc @dianakhuang 

@dyudyunov: can we make this toggleable? so we can turn it on and off?

dyudyunov commented:
hi @mariajgrimaldi 

> @dyudyunov: can we make this toggleable? so we can turn it on and off?

Sure we can, but I don't know if I'll have time to implement it till the end of May. 
I'll do my best though

mphilbrick211 commented:
Hi @dyudyunov! Following-up on this to see if you're able to push this forward?

dyudyunov commented:
> Hi @dyudyunov! Following-up on this to see if you're able to push this forward?

Hello @mphilbrick211 

I've lost this issue from my focus, thanks for the notification!

I'll try to update it by the end of the next week (I can't promise I can do it earlier)

dyudyunov committed:
feat: make default enrollment start togglable

dyudyunov committed:
fix: disable linting warnings for reimport

dyudyunov commented:
@mphilbrick211 @mariajgrimaldi 

I've added the feature toggle and made the new behavior switched off by default

To enable it - you'll need to update the cms settings.FEATURES:
`FEATURES['CREATE_COURSE_WITH_DEFAULT_ENROLLMENT_START_DATE'] = True` 
(the setting name is pretty long though self-descriptive enough IMHO)

Also rebased on master, so it's now ready for merge 🚀

mphilbrick211 commented:
Hi @mariajgrimaldi - just checking in on this!

mariajgrimaldi commented:
I'll review this during the week! Thanks

mariajgrimaldi commented:
Sorry - it's been crazy lately. This is on my review queue! Thanks for your patience.

dyudyunov commented:
It would be great to create a backport PR to Palm. I'm going to do so as soon as this one will be accepted

mariajgrimaldi reviewed (approved):
Thanks again for the contribution, @dyudyunov! I tested this again by ensuring the default enrollment start date was used as default if the flag was turned on. 

I'll post a message in the edx-platform CC slack channel so people know about this change! See you there.

robrap commented on xmodule/course_block.py:
```diff
@@ -57,6 +58,17 @@
+# .. toggle_implementation: SettingDictToggle
+# .. toggle_default: False
+# .. toggle_description: When enabled the newly created courses will have the enrollment_start_date
+#   set to DEFAULT_START_DATE.
```
Could you add something like: "The default behavior, when this is disabled, is ...".

dyudyunov replied:
Done, please check

robrap commented on xmodule/course_block.py:
```diff
@@ -57,6 +58,17 @@
+#   set to DEFAULT_START_DATE.
+# .. toggle_use_cases: open_edx
+# .. toggle_creation_date: 2023-06-22
+# .. toggle_target_removal_date: None
```
You should drop this annotation altogether.
```suggestion
```

In the future, another annotation will make it clear that this is a permanent toggle, but for now, you could something like: "This is intended to be a permanent option."

dyudyunov replied:
Replaced with the proposed description

robrap commented on xmodule/course_block.py:
```diff
@@ -57,6 +58,17 @@
 COURSE_VIDEO_SHARING_PER_VIDEO = 'per-video'
 COURSE_VIDEO_SHARING_ALL_VIDEOS = 'all-on'
 COURSE_VIDEO_SHARING_NONE = 'all-off'
+# .. toggle_name: FEATURES['CREATE_COURSE_WITH_DEFAULT_ENROLLMENT_START_DATE']
```
No change is being requested in this comment.

[inform] There has been an unofficial move away from using the `FEATURES` dict. We should probably have a DEPR to at least discuss it.
FYI: @feanil

robrap replied:
[inform] I created https://github.com/openedx/edx-platform/issues/33026 to discuss this topic. Again, no request for change.

robrap reviewed (commented):
Just added some comments about the toggle. Thank you.

dyudyunov committed:
docs: format toggle annotation

robrap commented on xmodule/course_block.py:
```diff
@@ -57,6 +58,18 @@
+# .. toggle_default: False
+# .. toggle_description: The default behavior, when this is disabled, is that newly created course has no
+#   enrollment_start date set. When the feature is enabled - the newly created courses will have the
+#   enrollment_start_date set to DEFAULT_START_DATE.
```
1. I fixed a typo and moved the the sentence about it being a permanent option to the description.
```suggestion
# .. toggle_description: The default behavior, when this is disabled, is that a newly created course has no
#   enrollment_start date set. When the feature is enabled - the newly created courses will have the
#   enrollment_start_date set to DEFAULT_START_DATE. This is intended to be a permanent option.
```

2. My understanding is that this toggle was added because some orgs (2U) did not want this toggled on. I do not know much about this, and if I were an operator, I would wonder why I should choose one option over the option:
  i. What is the consequence of having no start date set? Without knowing anything, that seems like it would be bad, but somehow it works for 2U. Why?
  ii. I think I read that DEFAULT_START_DATE is intended to be a date in the distant future so courses will not appear. Is that accurate, or only if you are using this toggle? Is the default used for other things? Maybe you could speak to that a bit more here?

Please update my first point, because it is simple. Regarding my second point, I am not the owner of this particular area of the code, so this is just a suggestion. You can decide how much you update the toggle docs, but I personally would rather longer docs that help me understand more completely what it is really for. That said, if you think most operators would know all of these things, maybe you've provided enough.

dyudyunov replied:
Thanks for the detailed review

> What is the consequence of having no start date set?

The toggle has no effect on the **start date** but only on the **enrollment start date**, but I've got your point, I'll think about how to describe the toggle behavior with more details

> I think I read that DEFAULT_START_DATE is intended to be a date in the distant future so courses will not appear. Is that accurate, or only if you are using this toggle? Is the default used for other things? Maybe you could speak to that a bit more here?

I just reused the DEFAULT_START_DATE constant. It was originally used only as a default value for the **course start** date and no changes were made there

robrap commented on xmodule/course_block.py:
```diff
@@ -57,6 +58,18 @@
+#   enrollment_start_date set to DEFAULT_START_DATE.
+# .. toggle_use_cases: open_edx
+# .. toggle_creation_date: 2023-06-22
+# .. toggle_target_removal_date: This is intended to be a permanent option.
```
Apologies that I wasn't clear. This annotation should not be used for permanent options. The sentence I added in the description is to help make the permanence of this toggle more explicit. 
```suggestion
```

In the future, https://github.com/openedx/edx-toggles/issues/284 proposes a`toggle_permanent_justification` annotation, that would make this explicit.

robrap reviewed (commented):
Thanks @dyudyunov. I added some comments about the toggle annotations. Hopefully they are clear. Please don't be blocked on any re-review by me unless you wish to. I am not adding an approval simply because I didn't review anything other than this annotation. Thanks for your updates.

dyudyunov committed:
docs: update toggle docstring

dyudyunov commented:
@mariajgrimaldi hi 👋

Is there something else blocking us from merging?)

mariajgrimaldi commented:
I'll be merging this now!

openedx-webhooks commented:
@dyudyunov 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/edx-platform&entry.867055334=https://github.com/openedx/edx-platform/pull/30954&entry.1484655318=https://github.com/dyudyunov&entry.752974735=2022-09-07+07:33&entry.1917517419=2023-08-22+14:42&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->

mariajgrimaldi commented:
@dyudyunov: let us know when you have the backport PR ready :)
